From 2006b1a874f6f6060a5f396d83ad827b13e78ca8 Mon Sep 17 00:00:00 2001
From: Nan Wu <wnan@google.com>
Date: Tue, 31 Mar 2026 18:33:32 -0700
Subject: [PATCH] Move intent redirect checks before intent resolution.

The checks for missing or invalid creator tokens and the extraction of intent creator information are now performed before calling supervisor.resolveIntent/resolveActivity. This ensures that the intent's component is not changed by resolveActivity method, which invalidates creatorToken.

Bug: 475542441
Test: Manually verify the PoC of the bug is fixed.
Flag: EXEMPT BUGFIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:93b9c2af63c6f161cd3287d27b892e0f2057c3ce
Merged-In: Iae6a7ef5cb569e1436c9a705b686488bf348b5ff
Change-Id: Iae6a7ef5cb569e1436c9a705b686488bf348b5ff
---
 .../server/wm/ActivityStartController.java    | 25 +++++++++++---
 .../android/server/wm/ActivityStarter.java    | 33 ++++++++++---------
 2 files changed, 37 insertions(+), 21 deletions(-)

diff --git a/services/core/java/com/android/server/wm/ActivityStartController.java b/services/core/java/com/android/server/wm/ActivityStartController.java
index 51025d204b46..b617638e0e16 100644
--- a/services/core/java/com/android/server/wm/ActivityStartController.java
+++ b/services/core/java/com/android/server/wm/ActivityStartController.java
@@ -451,13 +451,21 @@ public class ActivityStartController {
                 // Remove existing mismatch flag so it can be properly updated later
                 intent.removeExtendedFlags(Intent.EXTENDED_FLAG_FILTER_MISMATCH);
 
-                // Collect information about the target of the Intent.
-                ActivityInfo aInfo = mSupervisor.resolveActivity(intent, resolvedTypes[i],
-                        0 /* startFlags */, null /* profilerInfo */, userId, filterCallingUid,
-                        callingPid);
-                aInfo = mService.mAmInternal.getActivityInfoForUser(aInfo, userId);
                 int creatorUid = DEFAULT_INTENT_CREATOR_UID;
                 String creatorPackage = null;
+
+                // Check if the Intent was redirected
+                if ((intent.getExtendedFlags()
+                        & Intent.EXTENDED_FLAG_MISSING_CREATOR_OR_INVALID_TOKEN) != 0) {
+                    if (ActivityStarter.logAndAbortForIntentRedirect(mService.mContext,
+                            ActivityStarter.INTENT_REDIRECT_EXCEPTION_MISSING_OR_INVALID_TOKEN,
+                            intent, creatorUid, creatorPackage, filterCallingUid,
+                            callingPackage)) {
+                        Slog.d(TAG, "Not allowed to start activity since no intent creator token.");
+                        return START_CANCELED;
+                    }
+                }
+
                 if (ActivityManagerService.IntentCreatorToken.isValid(intent)) {
                     ActivityManagerService.IntentCreatorToken creatorToken =
                             (ActivityManagerService.IntentCreatorToken) intent.getCreatorToken();
@@ -468,6 +476,12 @@ public class ActivityStartController {
                     // leave creatorUid as -1 if the intent creator is the same as the launcher
                 }
 
+                // Collect information about the target of the Intent.
+                ActivityInfo aInfo = mSupervisor.resolveActivity(intent, resolvedTypes[i],
+                        0 /* startFlags */, null /* profilerInfo */, userId, filterCallingUid,
+                        callingPid);
+                aInfo = mService.mAmInternal.getActivityInfoForUser(aInfo, userId);
+
                 if (aInfo != null) {
                     try {
                         // Carefully collect grants without holding lock
@@ -498,6 +512,7 @@ public class ActivityStartController {
                                     callingPackage, securityException);
                         }
                     }
+
                     if ((aInfo.applicationInfo.privateFlags
                             & ApplicationInfo.PRIVATE_FLAG_CANT_SAVE_STATE) != 0) {
                         throw new IllegalArgumentException(
diff --git a/services/core/java/com/android/server/wm/ActivityStarter.java b/services/core/java/com/android/server/wm/ActivityStarter.java
index 7998bf56915a..741057202bc1 100644
--- a/services/core/java/com/android/server/wm/ActivityStarter.java
+++ b/services/core/java/com/android/server/wm/ActivityStarter.java
@@ -606,21 +606,6 @@ class ActivityStarter {
                 intent.setComponent(null /* component */);
             }
 
-            resolveInfo = supervisor.resolveIntent(intent, resolvedType, userId,
-                    0 /* matchFlags */,
-                    computeResolveFilterUid(callingUid, realCallingUid, filterCallingUid),
-                    realCallingPid);
-            if (resolveInfo == null) {
-                // Special case for profiles: If attempting to launch non-crypto aware app in a
-                // locked profile or launch an app in a profile that is stopped by quiet mode from
-                // an unlocked parent, allow it to resolve as user will be sent via confirm
-                // credentials to unlock the profile.
-                resolveInfo = resolveIntentForLockedOrStoppedProfiles(supervisor);
-            }
-
-            // Collect information about the target of the Intent.
-            activityInfo = supervisor.resolveActivity(intent, resolveInfo, startFlags,
-                    profilerInfo);
             // Check if the Intent was redirected
             if ((intent.getExtendedFlags() & Intent.EXTENDED_FLAG_MISSING_CREATOR_OR_INVALID_TOKEN)
                     != 0) {
@@ -637,6 +622,22 @@ class ActivityStarter {
                 }
                 // leave intentCreatorUid as -1 if the intent creator is the same as the launcher
             }
+
+            resolveInfo = supervisor.resolveIntent(intent, resolvedType, userId,
+                    0 /* matchFlags */,
+                    computeResolveFilterUid(callingUid, realCallingUid, filterCallingUid),
+                    realCallingPid);
+            if (resolveInfo == null) {
+                // Special case for profiles: If attempting to launch non-crypto aware app in a
+                // locked profile or launch an app in a profile that is stopped by quiet mode from
+                // an unlocked parent, allow it to resolve as user will be sent via confirm
+                // credentials to unlock the profile.
+                resolveInfo = resolveIntentForLockedOrStoppedProfiles(supervisor);
+            }
+
+            // Collect information about the target of the Intent.
+            activityInfo = supervisor.resolveActivity(intent, resolveInfo, startFlags,
+                    profilerInfo);
             // Carefully collect grants without holding lock
             if (activityInfo != null) {
                 if (android.security.Flags.contentUriPermissionApis()) {
@@ -3698,7 +3699,7 @@ class ActivityStarter {
         }
     }
 
-    private static boolean logAndAbortForIntentRedirect(@NonNull Context context,
+    static boolean logAndAbortForIntentRedirect(@NonNull Context context,
             @IntentRedirectErrorCode int errorCode, @NonNull Intent intent, int intentCreatorUid,
             @Nullable String intentCreatorPackage, int callingUid,
             @Nullable String callingPackage) {
-- 
2.55.0.897.gb25b4bd76c-goog

